### ISSUE TITLE
in Docker demo, dispatch /command "set current demo /increment" fails from non-tree server

### ISSUE THREAD

**Issue by gauge2** [MDSplus/mdsplus#2380]
For what I assume are obvious reasons, I have no problem running scripts from the tree-server where both shot.tcl and stop.tcl complete. I can also do that with python on the tree-server, but the moment I try to do it from another container, the error reported at the end above returns. 

I've ensured the mdsip servers are running using the mdsip.hosts file with `*    | hostuser` as the entry, mapping incoming users to the local user 'hostuser'. 

I've ensured that the mdsip servers are running on their respective ports. 

If I call `demo-mdstcl` and at the TCL prompt call the shot script `TCL>@/scripts/shot.tcl`, the output is 

``` 
TCL> @/scripts/shot.tcl 
TCL> set tree demo
TCL> set current demo /increment
Error: Unable to change current shot
Error message was: %SS-W-SUCCESS, Success
mdsdcl: --> failed on line 'set current demo /increment'
mdsdcl: --> failed on line '@/scripts/shot.tcl '
TCL>
```
Appears that one cannot increment the shot number.  

Why might that be? What other reasons would prevent the shot number from incrementing, raising exception, and stopping the server?

**Comment by zack-vii** [comment]
did you set the current shot at some point? increment requires an initial current shot number.. , e.g. set current 1. it could be as easy as that. otherwise, make sure that the current user (is it hostuser) has read-write permissions to shotid.sys in the tree_path, i.e. the paths pointed to by demo_path env.

**Comment by gauge2** [comment]
@zack-vii 

TL;DR: bc Docker runs `sh` as root user, when running `demo-mdstcl """@/scripts/create_demo_tree.tcl"""` as provided, creates the file as root and not hostuser so no-one can write to the tree from anywhere other than on the tree server. 

--- 
Reply:
Thanks for your reply, in `demo/scripts/shot.tcl`, it doesn't include a line to set the current shot. I tried setting it to the current shot before incrementing, but that did not allow me to increment.  I removed the set current /shot=0 and went to check permissions.

Looking at `shotid.sys` permissions, while the user was part of the group that owned the file, the file itself was owned by `root`. This is because the command in the `setup.sh` script provided in the demo calls to create the files from scratch (empty dir) which is:

```
demo-mdstcl """@/scripts/create_demo_tree.tcl"""
```

is just a wrapper to attach a shell to a plain old MDSplus:alpha container. Since operations within it are run as root, it creates the tree files as root and no one can write to it from any server bc hostuser does not have write permissions on `shotid.sys`.  

In the meantime, I added the following to the entrypoint for tree- and mdsip-server entrypoint scripts:

```
# Ensure mdsplus env vars are available to 'hostuser'
echo ". /etc/profile.d/mdsplus.sh" >>/home/hostuser/.bashrc
echo "export default_tree_path=tree_server::" >>/home/hostuser/.bashrc
```

This means that MDSplus env vars and the tree path are available to `hostuser`. 

Once this is done, I can open an interactive shell on an active tree-server container and create a tree. 
After this, I can run the rest of the sequence no problem. The only thing still not working is the `demo-dwscope` that is supposed to stream the event but all that is shown is the same "error connecting to database" message. 

So that clears that up. If that's all good, I'll close the issue since it's a permissions thing, but also the way the shot sequence in the demo as provided does not work with new trees (empty experimental tree directory ie. /trees/demo is empty) when creating demo tree.

**Comment by gauge2** [comment]
Error caused by not-having non-root write permissions when the tree is created, and if new shot taken by root user. Must act as non-root.
